sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -120,6 +148,129 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+    }
+
+    // Additional validation for splittable DoFn's.
+    if (processElement.isSplittable()) {
+      if (isBounded == null) {
+        isBounded =
+            processElement.hasReturnValue()
+                ? PCollection.IsBounded.UNBOUNDED
+                : PCollection.IsBounded.BOUNDED;
+      }
+      List<String> missingRequiredMethods = new ArrayList<>();
+      if (getInitialRestriction == null) {
+        missingRequiredMethods.add("@" + DoFn.GetInitialRestriction.class.getSimpleName());
+      }
+      if (newTracker == null) {
+        missingRequiredMethods.add("@" + DoFn.NewTracker.class.getSimpleName());
+      }
+      // @SplitRestriction and @GetRestrictionCoder are optional.
+      if (!missingRequiredMethods.isEmpty()) {
+        processElementErrors.throwIllegalArgument(
+            "Splittable, but does not define the following required methods: %s",
+            missingRequiredMethods);
+      }
+      processElementErrors.checkArgument(
+          processElement.trackerT().equals(newTracker.trackerT()),
+          "Has tracker type %s, but @%s method %s uses tracker type %s",
+          formatType(processElement.trackerT()),
+          DoFn.NewTracker.class.getSimpleName(),
+          format(newTrackerMethod),
+          formatType(newTracker.trackerT()));
+      getInitialRestrictionErrors.checkArgument(
+          getInitialRestriction.restrictionT().equals(newTracker.restrictionT()),
+          "Uses restriction type %s, but @%s method %s uses restriction type %s",
+          formatType(getInitialRestriction.restrictionT()),
+          DoFn.NewTracker.class.getSimpleName(),
+          format(newTrackerMethod),
+          formatType(newTracker.restrictionT()));
+      if (getRestrictionCoder != null) {
+        getInitialRestrictionErrors.checkArgument(
+            getRestrictionCoder
+                .coderT()
+                .isSubtypeOf(coderTypeOf(getInitialRestriction.restrictionT())),
+            "Uses restriction type %s, but @%s method %s returns %s "
+                + "which is not a subtype of %s",
+            formatType(getInitialRestriction.restrictionT()),
+            DoFn.GetRestrictionCoder.class.getSimpleName(),
+            format(getRestrictionCoderMethod),
+            formatType(getRestrictionCoder.coderT()),
+            formatType(coderTypeOf(getInitialRestriction.restrictionT())));
+      }
+      if (splitRestriction != null) {
+        getInitialRestrictionErrors.checkArgument(
+            splitRestriction.restrictionT().equals(getInitialRestriction.restrictionT()),
+            "Uses restriction type %s, but @%s method %s uses restriction type %s",
+            formatType(getInitialRestriction.restrictionT()),
+            DoFn.SplitRestriction.class.getSimpleName(),
+            format(splitRestrictionMethod),
+            formatType(splitRestriction.restrictionT()));
+      }
+    } else {
+      processElementErrors.checkArgument(
```

Comment:
Can you split this into two methods? `verifySplittableDoFn` and `verifyNonSplittableDoFn`? That would be better than having the two big if-blocks.